File: problems/LeetCode/binary_search/searchRange.py
def lower_bound (self, nums: List[int], target: int) -> List[int]:

		left, right = 0, len(nums) - 1

		while left <= right:

				mid = (left + right) // 2

				if target <= nums[mid]:
						right = mid - 1
				else:
						left = mid + 1

		return left

def searchRange(self, nums: List[int], target: int) -> List[int]:
		start = self.lower_bound(nums, target)
		if start == len(nums) or nums[start] != target:
				return [-1,-1]
		end = self.lower_bound(nums, target + 1) - 1
		return [start, end]

File: problems/LeetCode/TopInterview150/searchRange.ts
function searchRange(nums: number[], target: number): number[] {
  const start = bisectLeft(nums, target);

  if (start === nums.length || nums[start] !== target) return [-1, -1];

  const end = bisectLeft(nums, target + 1) - 1;
  return [start, end];
}

function bisectLeft(nums: number[], target: number): number {
  let lo = 0;
  let hi = nums.length;

  while (lo < hi) {
    const mid = lo + Math.floor((hi - lo) / 2);

    if (nums[mid] < target) {
      lo = mid + 1;
    } else {
      hi = mid;
    }
  }
  return lo;
}

function searchRange2(nums: number[], target: number): number[] {
  const start = lowerBound(nums, target);

  if (start === nums.length || nums[start] !== target) return [-1, -1];

  const end = higherBound(nums, target) - 1;
  return [start, end];
}

function lowerBound(nums: number[], target: number): number {
  let lo = 0;
  let hi = nums.length;

  while (lo < hi) {
    const mid = lo + Math.floor((hi - lo) / 2);

    if (nums[mid] < target) {
      lo = mid + 1;
    } else {
      hi = mid;
    }
  }
  return lo;
}

function higherBound(nums: number[], target: number): number {
  let lo = 0;
  let hi = nums.length;

  while (lo < hi) {
    const mid = lo + Math.floor((hi - lo) / 2);

    if (nums[mid] <= target) {
      lo = mid + 1;
    } else {
      hi = mid;
    }
  }
  return lo;
}

// Time: O(log(n))
// Space: O(1)
